# Multi-species homeodomain
This reposistory contains code to analyze single cell data for Xu et al. 2025. It also contatins a module to do a permutation test in support of figure 4a.

## Installation
- Install [micromamba](https://mamba.readthedocs.io/en/latest/installation/micromamba-installation.html)
- Clone the reposistory
```bash
git clone https://github.com/pnewstein/multi-species-homeodomain
```
- Create a new environment with installed dependencies
```bash
micromamba env create -f environment.yml
```

## Getting data

### Panther htdtf family
This is a list of panther families in the homeodomain class. This file is included in this repo but can be replicated with the following steps

1. https://www.pantherdb.org/panther/familyList.do?searchType=basic&fieldName=all&listType=6&fieldValue=PC00119
1. ensure that all items are displayed.
1. click send list to file
1. click save
1. take first word from each line that does not have a subfamily `awk '{print $1}' pantherGeneList.txt | grep -v : > panther_hdtf_families.txt`

### Adult female lamina neurons
This is the scRNAseq data generated by Xu et al., 2025
- H5AD can be downloaded from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM9179274
- The H5AD can be generated from the sequencing files in the same GEO reposistory with the following steps
1. Put .fastq.gz files into seq-data/raw-counts
1. Download 10x genomics cell ranger. 
1. Follow instructions to make a refrence genome from BDGP6.46
1. Invoke cellranger 
```bash
cellranger count --id=R1 --fastqs=../raw_counts --sample=Oregon-R1 --transcriptome=../Mkref/BDGP6_46  --localcores=28
cellranger count --id=R2 --fastqs=../raw_counts --sample=Oregon-R2 --transcriptome=../Mkref/BDGP6_46  --localcores=28
```
1. run `make data/adult_female_lamina_neurons.h5ad`  SAM, and scanpy (see Makefile for details)

### Bipolar neurons
- Download metadata and expression matrix from the [Single Cell Portal](https://singlecell.broadinstitute.org/single_cell/study/SCP3/retinal-bipolar-neuron-drop-seq#/)
- move both files into the seq-data directory
- run `make data.bpc.h5ad` to run scanpy (see Makefile for details).

### Other retinal cells
- AnnData can be downloaded from an h5ad file at https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE243413

## Usage
running `make` will make all of the images from the figures and do the permutation test 

### Making dotplots
All dotplots are made using src/make_dotplots.py

| Class                  | Image File           | H5ad file                             |
|------------------------|----------------------|---------------------------------------|
| Lamina                 | imgs/lamina_dots.svg | data/adult_female_lamina_neurons.h5ad |
| Bipolar Cells          | imgs/bpc_dots.svg    | data/bpc.h5ad                         |
| Amacine Cells          | imgs/AC.svg          | data/MRCA_full_normcounts.h5ad        |
| Horizonal Cells        | imgs/HC.svg          | data/MRCA_full_normcounts.h5ad        |
| Photoreceptors         | imgs/PR.svg          | data/MRCA_full_normcounts.h5ad        |
| Retinal Ganglion Cells | imgs/RGC.svg         | data/MRCA_full_normcounts.h5ad        |

### Permutation test
The Permutation test can be run from src/permute.py

